GH-133136: Revise QSBR to reduce excess memory held

nascheme opened the pull request:
This is a refinement of GH-135107.  Additional changes:

- track the size of the mimalloc pages that are deferred
- introduce `_Py_qsbr_advance_with_size()` to reduce duplicated code
- adjust the logic of when we advance the global write sequence and when we process the queue of deferred memory
- small fix for the goal returned in the advance case, it is safe to return the new global write sequence, not the next write sequence

With these changes, the memory held by QSBR is typically freed a bit more quickly and the process RSS stays a bit smaller.

Regarding the changes to advance and processing, GH-135107 has the following minor issues: if the memory threshold is exceeded when a new item is added, by `free_delayed()`, we immediately set `memory_deferred = 0` and process.  It is very unlikely that the goal has been reached for the newly added item.  If that's a big chunk of memory, we would have to wait until the *next* process in order to actually free it.  This PR tries to avoid that by storing the `seq` (local read sequence) as it was at last process time.  If that hasn't changed (this thread hasn't entered a quiescent state) then we wait before processing.  This at least gives a chance that other readers will catch up and the process can actually free things.

This PR also changes how often we can defer the advance of the global write sequence.  Previously, we deferred it up to 10 times.  However, I think there is not much benefit to advancing it unless we are nearly ready to process.  So, the `should_advance_qsbr()` is checking if it seems time to process.  The `_Py_qsbr_should_process()` checks if the local read sequence has been updated.  That means the write sequence has advanced (it's time to process) and the read sequence for this thread has also advanced.  This doesn't tell us that the other threads have advanced their read sequence but we don't want to pay the cost of checking that (would require "poll").

[pyperformance memory usage results](https://github.com/facebookexperimental/free-threading-benchmarking/blob/main/results/bm-20250612-3.15.0a0-3978e35-NOGIL/bm-20250612-vultr-x86_64-nascheme-gh_133136_qsbr_defer-3.15.0a0-3978e35-vs-base-mem.svg)

<!-- gh-issue-number: gh-133136 -->
* Issue: gh-133136
<!-- /gh-issue-number -->

colesbury committed:
gh-133136: Limit excess memory held by QSBR

The free threading build uses QSBR to delay the freeing of dictionary
keys and list arrays when the objects are accessed by multiple threads
in order to allow concurrent reads to proceeed with holding the object
lock. The requests are processed in batches to reduce execution
overhead, but for large memory blocks this can lead to excess memory
usage.

Take into account the size of the memory block when deciding when to
process QSBR requests.

colesbury committed:
Fix unused function warning

nascheme committed:
Re-work QSBR deferred advance and processing.

nascheme committed:
Update comments.

colesbury commented on Python/qsbr.c:
```diff
@@ -113,17 +118,47 @@ _Py_qsbr_advance(struct _qsbr_shared *shared)
     // about the wr_seq getting too far ahead of rd_seq, but if we ever use
     // 32-bit sequence numbers, we'll need to be more careful.
-    return _Py_atomic_add_uint64(&shared->wr_seq, QSBR_INCR) + QSBR_INCR;
+    return _Py_atomic_add_uint64(&shared->wr_seq, QSBR_INCR);
```
Why was this modified?

nascheme replied:
I think using "current value + QSBR_INCR" is unnecessarily conservative.  We have just advanced the write sequence.  If the caller of `free_deferred()` has removed the memory from the structure (which it should have), using the current (newly advanced) wr_seq value as the goal is safe.  That will free the memory a little faster.  This probably doesn't make a large difference but with the previous code we are waiting one extra advance and I don't think that's required in order to be safe.

colesbury replied:
`_Py_atomic_add_uint64` returns the *previous* value, not the result of the addition [^1], so this change means that it's using the old value of `wr_seq` as the goal

[^1]: This is to match the C11 behavior of `atomic_fetch_add`

nascheme replied:
Oh, that was kind of a serious bug.  I revert that.

colesbury commented on Objects/obmalloc.c:
```diff
@@ -1226,7 +1227,7 @@ _PyObject_XDecRefDelayed(PyObject *ptr)
     assert(!((uintptr_t)ptr & 0x01));
     if (ptr != NULL) {
-        free_delayed(((uintptr_t)ptr)|0x01);
+        free_delayed(((uintptr_t)ptr)|0x01, 64);
```
In practice, I don't think there's a significant difference between `0` and `64` (or any small number) here. I had just stuck in 64 as a placeholder.

Maybe we should just use `0` for now and add a comment?

In practice, we only use this for `_PyObject_SetDict()` currently. I think the choice is basically use a small number (that effectively behaves like zero) or a large number to force processing. (We could also push this choice up to the caller of `_PyObject_XDecRefDelayed()`).

nascheme replied:
Okay.

colesbury commented on Objects/obmalloc.c:
```diff
@@ -1299,8 +1300,11 @@ maybe_process_interp_queue(struct _Py_mem_interp_free_queue *queue,
     _PyThreadStateImpl *tstate_impl = (_PyThreadStateImpl *)tstate;
+    if (!_Py_qsbr_should_process(tstate_impl->qsbr)) {
+        return;
+    }
```
I'd prefer this to be pulled up into the caller. We only have one caller currently, so it doesn't make a logical difference, but I'd prefer to pull the conditional logic up higher:

1) Symmetry between `_PyMem_ProcessDelayed` and `_PyMem_ProcessDelayedNoDealloc`: both actually process the memory to be freed
2) I think these sorts of heuristics tend to be sensitive to the context and `free_delayed` has more information about when it's called.

nascheme replied:
Okay.

colesbury commented on Python/qsbr.c:
```diff
@@ -113,17 +118,47 @@ _Py_qsbr_advance(struct _qsbr_shared *shared)
 
 uint64_t
-_Py_qsbr_deferred_advance(struct _qsbr_thread_state *qsbr)
+_Py_qsbr_advance_with_size(struct _qsbr_thread_state *qsbr, size_t size)
```
I'd like to keep `deferred` in the function name if it isn't guaranteed to advance the shared sequence counter.

nascheme replied:
Okay.  I've used `_Py_qsbr_deferred_advance_for_page()` and `_Py_qsbr_deferred_advance_for_free()`.  Long but hopefully clear.

colesbury commented on Objects/obmalloc.c:
```diff
@@ -139,7 +139,10 @@ _PyMem_mi_page_maybe_free(mi_page_t *page, mi_page_queue_t *pq, bool force)
-        page->qsbr_goal = _Py_qsbr_deferred_advance(tstate->qsbr);
+
+        size_t bsize = mi_page_block_size(page);
+        page->qsbr_goal = _Py_qsbr_advance_with_size(tstate->qsbr, page->capacity*bsize);
```
This might be the right heuristic, but this is a bit different from `_PyMem_FreeDelayed`:

1) `_PyMem_FreeDelayed` holds onto the memory until quiescence. It prevents the memory from being used for any purpose.
2) `_PyMem_mi_page_maybe_free` only prevents the page from being used by another thread or for a different size class. That's a lot less restrictive.

nascheme replied:
Ah good point.  The memory being held (avoiding collection) by mimalloc is not at all the same as the deferred frees.   I reworked the PR so that memory is tracked separately.  I also decoupled the write sequence advance from the triggering of `_PyMem_ProcessDelayed()`, used `process_seq` as a target value for the read sequence.  

Now `_qsbr_thread_state` is larger than 64 bytes.  I don't think that should be a problem.

nascheme committed:
Revise based on review feedback.

* Keep separate count of mimalloc page memory that is deferred from
  collection.  This memory doesn't get freed by _PyMem_ProcessDelayed().
  We want to advance the write sequence if there is too much of it
  but calling _PyMem_ProcessDelayed() is not helpful.

* Use `process_seq` variable to schedule the next call to
  `_PyMem_ProcessDelayed()`.

* Rename advance functions to have "deferred" in name.

* Move `_Py_qsbr_should_process()` call up one level.

nascheme committed:
Revert change to _Py_qsbr_advance().

Since _Py_atomic_add_uint64() returns the old value, we need to add
QSBR_INCR.

nascheme committed:
Call _PyMem_FreeDelayed() from eval breaker.

Refactor code to keep obmalloc logic out of the qsbr.c file.  Call
_PyMem_ProcessDelayed() from the eval breaker.

nascheme committed:
Merge branch 'origin/main' into gh-133136-qsbr-defer-process

colesbury commented on Objects/obmalloc.c:
```diff
@@ -1141,8 +1171,38 @@ free_work_item(uintptr_t ptr, delayed_dealloc_cb cb, void *state)
+should_advance_qsbr_for_free(struct _qsbr_thread_state *qsbr, size_t size)
+{
+    qsbr->deferred_count++;
+    qsbr->deferred_memory += size;
```
There's a potential overflow here.

In my original PR, I first checked if `size > QSBR_FREE_MEM_LIMIT` to avoid that sort of potential problem

colesbury commented on Objects/obmalloc.c:
```diff
@@ -124,6 +124,29 @@ _PyMem_mi_page_is_safe_to_free(mi_page_t *page)
+{
+    size_t bsize = mi_page_block_size(page);
+    size_t page_size = page->capacity*bsize;
+    qsbr->deferred_page_memory += page_size;
```
Potential overflow here

colesbury reviewed (approved):
Looks good to me. Two comments below

nascheme committed:
Merge 'origin/main' into gh-133136-qsbr-defer-process

nascheme committed:
Avoid overflow in deferred memory accumulators.
